### PR TITLE
Add interpreter for CholeskyOp

### LINKED ISSUE #1123: Add interpreter for CholeskyOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
Here are the constraints for CholeskyOp:
```
(I1) `a` is a tensor of floating-point or complex type.
(I2) `lower` is a tensor constant of `i1` type.
(C1) `a` and `result` have the same type.
(C2) rank(`a`) >= 2.
(C3) dim(`a`, -2) = dim(`a`, -1).
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) `a` is not a tensor of floating-point type or complex type. (Covered by ODS).
I2: a) `lower` is not a tensor constant of `i1` type. (Covered by ODS).
C1: a) type(a) != type(result).
C2: a) rank(a) < 2.
C3: a) dim(`a`, -2) != dim(`a`, -1).
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
C1a: type(a) != type(result).
C2a: rank(a) < 2.
C3a: dim(`a`, -2) != dim(`a`, -1).
```

Notes:
* Implementation inspired by the [Cholesky–Banachiewicz algorithm](https://en.wikipedia.org/wiki/Cholesky_decomposition).

closes #1123

### CODE REVIEW

**Comment by ghpvnist** [comment]
Naming terminology isn't so great, but feel free to suggest an alternative.

File: stablehlo/reference/Ops.cpp [hunk 1]
```diff
@@ -430,6 +435,83 @@ Tensor evalCeilOp(const Tensor &operand, ShapedType resultType) {
...
+  if (a.getRank() == 2) return cholesky(a);
+
+  auto getScalarTensor = [&](auto value) {
+    return makeTensor(DenseElementsAttr::get(
+        RankedTensorType::get({},
+                              IntegerType::get(a.getType().getContext(), 64)),
+        {value}));
+  };
+
+  SmallVector<int64_t> aSliceShape(aShape.end() - 2, aShape.end());
```
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:486]: aSliceShape -> aNonBatchShape

File: stablehlo/reference/Ops.cpp [hunk 2]
```diff
@@ -430,6 +435,83 @@ Tensor evalCeilOp(const Tensor &operand, ShapedType resultType) {
...
+  auto batchingSizes =
+      Sizes(SmallVector<int64_t>(aShape.begin(), aShape.end() - 2));
+  for (auto batchIt =
+           IndexSpaceIterator(batchingSizes, Sizes(a.getRank() - 2, 0));
+       batchIt != IndexSpaceIterator(batchingSizes, std::nullopt); ++batchIt) {
+    SmallVector<Tensor> startIndices;
+    for (auto index : *batchIt) startIndices.push_back(getScalarTensor(index));
+    startIndices.append({getScalarTensor(0L), getScalarTensor(0L)});
+
+    auto aReshaped = Sizes(a.getRank() - 2, 1);
```
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:496]: aReshape -> sliceSizes

File: stablehlo/tests/ops_stablehlo.mlir [hunk 3]
```diff
@@ -1420,34 +1420,18 @@ func.func @cholesky(%arg0: tensor<1x2x2xf32>) -> tensor<1x2x2xf32> {
...
 
-func.func @cholesky_invalid_elt(%arg0: tensor<1x2x2xi32>) -> tensor<1x2x2xi32> {
-  // expected-error@+1 {{operand #0 must be tensor of f8E4M3FN type or f8E5M2 type or f8E4M3FNUZ type or f8E5M2FNUZ type or 16-bit float or 32-bit float or 64-bit float or bfloat16 type or complex type with 32-bit float or 64-bit float elements values, but got 'tensor<1x2x2xi32>}}
-  %0 = "stablehlo.cholesky"(%arg0) { lower = true } : (tensor<1x2x2xi32>) -> tensor<1x2x2xi32>
-  func.return %0: tensor<1x2x2xi32>
-}
-
-// -----
-
-func.func @cholesky_wrong_infer_shape(%arg0: tensor<1x2x2xf32>) -> tensor<1x2x2x2xf32> {
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/tests/ops_stablehlo.mlir:1447]: Yes we do. As per the [checklist](https://github.com/openxla/stablehlo/blob/main/docs/reference_checklist.md) no. 5, we only require one positive test, which [infer_stablehlo.mlir](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/infer_stablehlo.mlir#L84-L90) already has.

File: stablehlo/reference/Ops.cpp [hunk 4]
```diff
@@ -439,6 +444,83 @@ Tensor evalCeilOp(const Tensor &operand, ShapedType resultType) {
...
+  if (a.getRank() == 2) return cholesky(a);
+
+  auto getScalarTensor = [&](auto value) {
+    return makeTensor(DenseElementsAttr::get(
+        RankedTensorType::get({},
+                              IntegerType::get(a.getType().getContext(), 64)),
+        {value}));
+  };
+
+  SmallVector<int64_t> aNonBatchShape(aShape.end() - 2, aShape.end());
```
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:495]: Can we make `aNonBatchShape` -> `nonBatchingSizes`? This was inspired from the very next statement. I believe they should be named in a way so that they seem related.

File: stablehlo/tests/ops_stablehlo.mlir (hunk 3 above)
[the remark this answers is not in the document]
Comment (sdasgup3) [reply stablehlo/tests/ops_stablehlo.mlir:1447]: maybe I dd not get this. Shouldn't we need a negative test other than the positive tests mentioned below
```
5(i)
Make sure that there is at least one test for each constraint in the verifier and type inference methods; constraints covered in ODS will not be tested. These tests will mostly be negative tests exercising the fact that the constraints are not met. Make sure to add at least one positive test in case there are none.
```

[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/tests/ops_stablehlo.mlir:1447]:
```
Make sure that there is at least one test for each constraint in the verifier
and type inference methods
```

The one test mentioned aboce would be a positive test which

```
Make sure to add at least one positive test in case there are none.
```

this statement says to add if there are none. Since we have one positive test, we've satisfied that there is at least one test for type inference.

```
These tests will mostly be negative tests exercising the fact that the constraints are not met. 
```

This just states that in most cases the tests would be negative (constraint tests) and we could have positive (type inference) tests.

[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/tests/ops_stablehlo.mlir:1447]: I think this requires rewording as it is not obvious and unnecessarily tricky. Will add a PR to address this.

**Review by sdasgup3 — APPROVED** [review]

**Review by burmako — APPROVED** [review]
Thank you for this contribution! This is clearly a non-trivial amount of work.

As discussed earlier today, [HLO evaluator](https://github.com/tensorflow/tensorflow/blob/f7b6dcfb9eb05e37c6738d4007b3f3c510da6784/tensorflow/compiler/xla/hlo/evaluator/hlo_evaluator.cc) employes a different strategy for providing support for ops like kCholesky. Instead of implementing their evaluation, it relies on expander passes, e.g. [cholesky_expander.h](https://github.com/tensorflow/tensorflow/blob/f7b6dcfb9eb05e37c6738d4007b3f3c510da6784/tensorflow/compiler/xla/service/cholesky_expander.h), which decompose these ops into other HLO ops which are then evaluated.

I think we should explore using this strategy for BatchNorm*Op, CholeskyOp, RngBitGeneratorOp, RngOp and TriangularSolveOp. We cannot use HLO passes in this repository, so this will not be super straightforward, but figuring this out will likely also synergize with the needs of the OpenXLA compiler, so I think it will likely be worth it.

I don't think this should block merging already existing PRs for BatchNorm*Ops and for CholeskyOp, but let's put a pause on implementing other ops mentioned above (plus, FftOp, whose implementation is also quite nontrivial) until deciding on this.
